src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs

```diff
@@ -219,14 +228,27 @@ public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedF
                     Convert.FromBase64String(s),
                     int.Parse(i));
 
-                var clientKey = _hmac(encoding, saltedPassword, "Client Key");
+                byte[] clientKey;
+                byte[] serverKey;
+                var cacheKey = new CacheKey(saltedPassword, Convert.FromBase64String(s), int.Parse(i));
+
+                if (_cache.Get(cacheKey) != null)
```

Comment:
If we use TryGet and strong typing we can use:

    if (_cache.TryGet(cacheKey, out var cacheEntry))
    {
        clientKey = cacheEntry.ClientKey;
        serverKey = cacheEntry.ServerKey;
    }
    else
    {
        // ...
    }